### ISSUE TITLE
LDAP support in negotiate authentication fails with "connection reset by peer"

### ISSUE THREAD

**Issue by ramon-garcia** [dotnet/aspnetcore#27968]
When using LDAP under Linux, so that one can have group membership in negotiate authentication, the connection to the LDAP server (domain controller) fails with error  "connection reset by peer".

The reason is that the connection to the LDAP server is made when the server application starts up. But if the connection is idle, it is closed. Thus the error when an query is made.

Either the connection should be retried after receiving ECONNRESET error, or tcp keepalive should be enabled.

In addition, the connection is unencrypted, therefore ldap user and password are sent in cleartext. It should be possible to enable TLS in LdapSettings.

Of course, the right thing to do would be to parse the group membership data from the Kerberos ticket.

**Comment by blowdart** [comment]
LDAP being unencrypted under Linux is a limitation of the libraries we're using, and isn't going to change any time soon I'm afraid.

**Comment by Tratcher** [comment]
> Of course, the right thing to do would be to parse the group membership data from the Kerberos ticket.

The APIs we use don't give us access to the raw kerberos ticket or role data, and the ticket doesn't always contain the roles. Hence the LDAP lookup.

**Comment by ramon-garcia** [comment]
> > Of course, the right thing to do would be to parse the group membership data from the Kerberos ticket.
> 
> The APIs we use don't give us access to the raw kerberos ticket or role data, and the ticket doesn't always contain the roles. Hence the LDAP lookup.

Yes we can have access to the PAC authorization data, with the function gss_get_name_attribute with attribute name "urn:mspac:". For instance, the [Samba code does that](https://git.samba.org/?p=samba.git;a=blob;f=auth/kerberos/gssapi_pac.c).

This is the solution with MIT Kerberos. (Heimdal Kerberos is better at many things, and had integrated PAC support earlier, but for now let us stay focused in solving the issue)

**Comment by blowdart** [comment]
As @Tratcher said the APIs available to ASP Net don't get access to the raw ticket, or the PAC data, so we can't parse it. So, that is not an option for us.

**Comment by JunTaoLuo** [comment]
Need to keep an eye out on https://github.com/dotnet/runtime/issues/43890.
